enhancements/update/targeted-update-edge-blocking.md

```diff
@@ -0,0 +1,725 @@
+// conditionally recommended for this cluster. Consumers interested
+// in the set of updates that are actually recommended for this
+// cluster should use availableUpdates. This list may be empty if no
+// updates are recommended, if the update service is unavailable, or
+// if an empty or invalid channel has been specified.
+// +optional
+conditionalUpdates []ConditionalUpdate `json:"conditionalUpdates,omitempty"`
+```
+
+The `availableUpdates` documentation will be adjusted to read:
+
+```go
+// availableUpdates contains the subset of conditionalUpdates that
+// apply to this cluster.  Updates which appear in conditionalUpdates
+// but not in availableUpdates may expose this cluster to known
+// issues.
+```
+
+The new ConditionalUpdate type will have the following schema:
+
+```go
+// ConditionalUpdate represents an update which is recommended to some
+// clusters on the version the current cluster is reconciling, but which
+// may not be recommended for the current cluster.
+// +k8s:deepcopy-gen=true
+type ConditionalUpdate struct {
+	// release is the target of the update.
+	// +required
+	release Release `json:"release"`
+
+	// blockedClusters represents the subset of clusters for which the conditional update is not recommended.
+	// +required
+	BlockedClusters []ConditionalUpdateBlocker `json:"blockedClusters"`
+
+	// conditions represents the observations of the conditional update's
+	// current status. Known types are:
+	// * Recommended, for whether the update is recommended for the current cluster.
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
+}
+```
+
+The new ConditionalUpdateBlocker type will have the following schema:
+
+```go
+// ConditionalUpdateBlocker represents a reason and cluster-set for
+// blocking a conditional update.
+// +k8s:deepcopy-gen=true
+type ConditionalUpdateBlocker struct {
+	// url contains information about this release. This URL is set by
+	// the 'url' metadata property on a release or the metadata returned by
+	// the update API and should be displayed as a link in user
+	// interfaces. The URL field may not be set for test or nightly
+	// releases.
+	// +optional
+	URL URL `json:"url,omitempty"`
+
+	// reason is the CamelCase reason for the condition's current status.
+	// +optional
+	Reason string `json:"reason,omitempty"`
+
+	// message provides additional information about the current condition.
+	// This is only to be consumed by humans. It may contain Line Feed
+	// characters (U+000A), which should be rendered as new lines.
+	// +optional
+	Message string `json:"message,omitempty"`
+
+	// promQL is a PromQL query describing affected clusters. The query
+	// should return a 0 if the update should be allowed and a 1 if the
+	// update should be blocked.
+	// +optional
+	PromQL string `json:"promQL,omitempty"`
```

Comment:
if I wanted to restrict on other criteria in the future (or even now), having this be a discriminated union could have value.  Platform is clear (you can select this in promql), but restricting based on "this is configured in your cluster" looks valuable.

Reply:
my concern w/ an api like that is we need to decide up front that we will only ever want to set one of those fields at a time for a given edge (vs wanting to AND or OR the fields together to make a decision about recommendation).

Reply:
That's fair; at some point we would risk creating our own expression language. Although we might need to, for those cases where prometheus isn't present, and in that case I would expect the `Type` to be the name of that language and for the expression to go into the `Query` field.

Reply:
Ok, 6330580 -> 32fb747 pivots to:

```yaml
when:
- type: promql
  promql: |
    0 * cluster_infrastructure_provider{type=~"VSphere|None"}
    or
    cluster_infrastructure_provider
- type: platform
  sadPlatforms:
  - VSphere
  - None
```

which should be the "discriminated union" pattern discussed above.  There's a bunch of hoop-jumping  to work around golang/go#6213; let me know if folks prefer a different approach to that, like:

```yaml
when:
- type: promql
  properties:
    promql: ...